addons/api/addon/workers/sqlite-worker.js

```diff
@@ -6,21 +6,155 @@
+        }
+      }
+
+      // If we have more than 20 files, that might indicate a problem so let's just
+      // remove the VFS and wipe all the client data across users and restart.
+      if (poolUtil.getCapacity() > 20) {
+        await poolUtil.removeVfs();
+      }
+    }
+  },
+  clearDatabase: () => {
+    db.exec(CLEAR_DB);
+    db.exec(CREATE_TABLES(SCHEMA_VERSION));
+  },
+  deleteDatabase: () => {
+    const name = db.dbFilename();
+    if (db) {
+      db.close();
+      db = null;
+    }
+
+    poolUtil.unlink(`/${name}`);
+  },
+  downloadDatabase: () => {
+    return sqlite3.capi.sqlite3_js_db_export(db);
+  },
+  fetchResource: ({ sql, parameters = [] }) => {
     try {
-      const sqlite3 = await sqlite3InitModule({
-        print: console.log,
-        printErr: console.error,
+      return db.exec({
+        sql,
+        bind: parameters,
+        rowMode: 'object',
       });
-      console.log(`SQLite Version: ${sqlite3?.version.libVersion}`);
-      const poolUtil = await sqlite3.installOpfsSAHPoolVfs();
-      new poolUtil.OpfsSAHPoolDb('/boundary');
     } catch (err) {
-      console.error('Initialization error:', err.name, err.message);
+      // TODO: Add error handling to rest of methods and properly return the error back to caller
+      console.error('Fetch resource error:', err, sql, parameters);
+      throw err;
     }
   },
+  insertResource: ({ resource, items }) => {
+    if (items.length === 0) {
+      return [];
+    }
+    const insertStatementGenerator = INSERT_STATEMENTS[resource];
+
+    if (!insertStatementGenerator) {
+      throw new Error(`Insert statement for resource "${resource}" not found.`);
+    }
+
+    const numberOfParameters = items.reduce((acc, item) => {
+      return acc + item.length;
+    }, 0);
+
+    // We want to use as many VALUES as part of the insert statement as it's quicker
+    // than doing multiple inserts. If we have too many rows to insert, we'll manually
+    // chunk them up to avoid hitting the maximum number of host parameters
+    // and execute these separate insert statements in a transaction
+    if (numberOfParameters > MAX_HOST_PARAMETERS) {
+      const chunkSize = Math.floor(MAX_HOST_PARAMETERS / items[0].length);
+
+      return db.transaction(() => {
+        const results = [];
+        for (let i = 0; i < items.length; i += chunkSize) {
+          const chunk = items.slice(i, i + chunkSize);
+          const result = db.exec({
+            sql: insertStatementGenerator(chunk),
+            bind: chunk.flat(),
+            rowMode: 'object',
+          });
+          results.push(result);
+        }
+        return results;
+      });
+    }
+
+    db.exec({
+      sql: insertStatementGenerator(items),
+      bind: items.flat(),
+      rowMode: 'object',
+    });
+  },
+  deleteResource: ({ resource, ids }) => {
+    if (ids.length === 0) {
+      return [];
+    }
+
+    // TODO: Test with a large amount of deletions, should we chunk here as well?
```

Comment:
None blocking comment: at first read of this comment I couldn't think of a case with such amount of deletions that will require us to chunk them, but giving a second though I might be missing context.

What cases do you think might require us to chunk deletions?

Perhaps a user of a company with big infra that has a quite outdated DB in memory? I guess that will require quite a bit of deletions of the old data and insertions of the new one.

In any case, IMHO no need to change this within this PR, we can take care of this later during the implementation phase.

Reply:
Yeah exactly, that's why I didn't do anything yet since it's also a pain to test to have that many deleted resources, but it would basically be if for whatever reason they deleted more than 32,766 resources in between refreshes.

Reply:
Thanks for the explanation

Reply:
I added a test for this and it did fail so I fixed the issue